scaladoc/src/dotty/tools/scaladoc/snippets/SnippetCompiler.scala

```diff
@@ -0,0 +1,110 @@
+package dotty.tools.scaladoc
+package snippets
+
+import dotty.tools.io.{AbstractFile, VirtualDirectory}
+import dotty.tools.dotc.Driver
+import dotty.tools.dotc.core.Contexts.Context
+import dotty.tools.dotc.core.Mode
+import dotty.tools.dotc.config.Settings.Setting._
+import dotty.tools.dotc.interfaces.SourcePosition
+import dotty.tools.dotc.ast.Trees.Tree
+import dotty.tools.dotc.interfaces.{SourceFile => ISourceFile}
+import dotty.tools.dotc.reporting.{ Diagnostic, StoreReporter }
+import dotty.tools.dotc.parsing.Parsers.Parser
+import dotty.tools.dotc.{ Compiler, Run }
+import dotty.tools.io.{AbstractFile, VirtualDirectory}
+import dotty.tools.repl.AbstractFileClassLoader
+import dotty.tools.dotc.util.SourceFile
+import dotty.tools.dotc.interfaces.Diagnostic._
+
+import scala.util.{ Try, Success, Failure }
+
+class SnippetCompiler(
+  val snippetCompilerSettings: Seq[SnippetCompilerSetting[_]],
+  val scalacOptions: String = "",
+  target: AbstractFile = new VirtualDirectory("(memory)")
+):
+  object SnippetDriver extends Driver:
+    val currentCtx =
+      val rootCtx = initCtx.fresh.addMode(Mode.ReadPositions).addMode(Mode.Interactive)
+      rootCtx.setSetting(rootCtx.settings.YretainTrees, true)
+      rootCtx.setSetting(rootCtx.settings.YcookComments, true)
+      rootCtx.setSetting(rootCtx.settings.YreadComments, true)
+      val defaultFlags =
+        List("-color:never", "-unchecked", "-deprecation", "-Ximport-suggestion-timeout", "0")
```

Comment:
Some settings are set with `setSetting` and some are strings passed to `setup`, it would be nicer and more type-safe to always use `setSetting`:
```scala
rootCtx.setSetting(rootCtx.settings.color, "never")
  .setSetting(...) // etc
```
Also it seems weird to explicitly enable `-unchecked` and `-deprecation` here, why those settings in particular? If the user wants them they can add them in `scalacOptions` themselves.

Reply:
The reason behind exactly these flags is that mdoc uses them. But I think it's OK to remove them and let user choose.